File: cash.c
//Calcuate fewest coins required for change with a greedy algorithm
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int main(void)
{
    //Require users change input to be a positive float
    float change;
    do
    {
        change = get_float("Change owed: ");
    }
    while (change < 0);

    //Convert change to integer to avoid imprecisions
    change *= 100;
    change = round(change);
    int cents = change;

    //Determine number of coins by taking largest coin possible (greedy)
    int num_coins = 0;
    while (cents > 0)
    {
        if (cents >= 25)
        {
            cents -= 25;
            num_coins++;
        }
        else if (cents >= 10)
        {
            cents -= 10;
            num_coins++;
        }
        else if (cents >= 5)
        {
            cents -= 5;
            num_coins++;
        }
        else
        {
            cents -= 1;
            num_coins++;
        }
    }

    //Print total number of coins
    printf("%i\n", num_coins);
}

File: cash.py
"""Cash homework problem"""
from cs50 import get_float


def main():
    """Use a greedy algorithm to obtain the fewest number of coins required for change specified."""

    # Obtain change amount from user
    change = get_float("Change owed: ")

    # Ensure input is positive
    while change < 0:
        change = get_float("Change owed: ")

    # Convert change to cent and round to 2 d.p.
    change *= 100
    change = round(change)

    # Initiate variable for number of coins and list of coin values
    coins = 0
    values = [25, 10, 5, 1]

    # Iterate over coin values
    for coin in values:

        # Determine number of coins used and change left for each coin
        coins += change // coin
        change = change % coin

    # Print number of coins used
    print(f"{coins}")


if __name__ == "__main__":
    # If program is run call main()
    main()